src/init.cpp

```diff
@@ -1686,26 +1686,24 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 break;
             }
 
-            bool failed_rewind{false};
-            // Can't hold cs_main while calling RewindBlockIndex, so retrieve the relevant
-            // chainstates beforehand.
-            for (CChainState* chainstate : WITH_LOCK(::cs_main, return chainman.GetAll())) {
-                if (!fReset) {
-                    // Note that RewindBlockIndex MUST run even if we're about to -reindex-chainstate.
-                    // It both disconnects blocks based on the chainstate, and drops block data in
-                    // BlockIndex() based on lack of available witness data.
-                    uiInterface.InitMessage(_("Rewinding blocks...").translated);
-                    if (!chainstate->RewindBlockIndex(chainparams)) {
-                        strLoadError = _(
-                            "Unable to rewind the database to a pre-fork state. "
-                            "You will need to redownload the blockchain");
-                        failed_rewind = true;
-                        break; // out of the per-chainstate loop
+            bool needs_ibd{false};
+            {
```

Comment:
We can avoid some deep nesting and the need for a temporary variable to pass the result of the predicate out of the loop by using an stl algorithm:

```diff
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1686,27 +1686,19 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 break;
             }
 
-            bool needs_ibd{false};
-            {
-                LOCK(cs_main);
-                for (CChainState* chainstate : chainman.GetAll()) {
-                    if (!fReset) {
-                        uiInterface.InitMessage(_("Checking if IBD is needed...").translated);
-                        if (chainstate->NeedsIBD(chainparams)) {
-                            strLoadError = _(
-                                "Segwit blocks are insufficiently validated. "
-                                "Please delete blocks dir and chain state dir and restart.");
-                            needs_ibd = true;
-                            break; // out of the per-chainstate loop
-                        }
-                    }
+            if (!fReset) {
+                uiInterface.InitMessage(_("Checking if redownload is needed...").translated);
+                LOCK(cs_main);
+                auto chainstates{chainman.GetAll()};
+                if (std::any_of(chainstates.begin(), chainstates.end(),
+                                [&chainparams] (CChainState* cs) {return cs->NeedsIBD(chainparams);})) {
+                    strLoadError = _(
+                        "Segwit blocks are insufficiently validated. "
+                        "Please delete blocks dir and chain state dir and restart.");
+                    break;
                 }
             }
 
-            if (needs_ibd) {
-                break; // out of the chainstate activation do-while
-            }
-
             bool failed_verification = false;
```

This is exactly what `std::any_of` is designed for. Instead of creating a boolean, manually iterating through a range, executing a predicate on each member and then setting the boolean and breaking out if the predicate returns true, we can let the standard library do the heavy lifting. That makes the code more expressive by raising the level of abstraction and communicating what we want the code to do, and not what the individual steps to do it are.

Reply:
Beautiful change. Thank you for teaching so patiently!

Reply:
Ooops. I guess the lambda needs an annotation too. Something like this maybe:

```
                if (std::any_of(chainstates.begin(), chainstates.end(),
                                [&chainparams] (CChainState* cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {return cs->NeedsIBD(chainparams);})) {
```

Reply:
That worked. For my education, what happened there and how did you know the lambda annotation would help?

I suspect the annotation is evaluated with the limited scope of the bound variables, so it does not "know" about the locks that are acquired in the outer scope. Is that even close?

Reply:
As discussed, there was an error message about `NeedsIBD()` requiring cs_main. Since the compiler does not know that when it's inside the lambda, it's in a scope that already has cs_main, we need to annotate it.